test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs

```diff
@@ -0,0 +1,868 @@
+            {
+                new TestBinderTypeProvider(),
+                new TestBinderTypeProvider(),
+            };
+
+            var provider = CreateProvider(binderProviders);
+
+            // Act
+            var metadata = provider.GetMetadataForType(typeof(string));
+
+            // Assert
+            Assert.Null(metadata.BinderType);
+        }
+
+        [Fact]
+        public void BinderTypeProviders_Fallback()
+        {
+            // Arrange
+            var attributes = new[]
+            {
+                new TestBinderTypeProvider(),
+                new TestBinderTypeProvider() { BinderType = typeof(string) }
+            };
+
+            var provider = CreateProvider(attributes);
+
+            // Act
+            var metadata = provider.GetMetadataForType(typeof(string));
+
+            // Assert
+            Assert.Same(typeof(string), metadata.BinderType);
+        }
+
+        [Fact]
+        public void BinderTypeProviders_FirstAttributeHasPrecedence()
+        {
+            // Arrange
+            var attributes = new[]
+            {
+                new TestBinderTypeProvider() { BinderType = typeof(int) },
+                new TestBinderTypeProvider() { BinderType = typeof(string) }
+            };
+
+            var provider = CreateProvider(attributes);
+
+            // Act
+            var metadata = provider.GetMetadataForType(typeof(string));
+
+            // Assert
+            Assert.Same(typeof(int), metadata.BinderType);
+        }
+
+        private IModelMetadataProvider CreateProvider(params object[] attributes)
+        {
+            return new AttributeInjectModelMetadataProvider(attributes);
+        }
+
+        private class TestBinderTypeProvider : IBinderTypeProviderMetadata
+        {
+            public Type BinderType { get; set; }
+
+            public BindingSource BindingSource { get; set; }
+        }
+
+        private class DataTypeWithCustomDisplayFormat : DataTypeAttribute
+        {
+            public DataTypeWithCustomDisplayFormat() : base("Custom datatype")
+            {
+                DisplayFormat = new DisplayFormatAttribute
+                {
+                    DataFormatString = "value",
+                };
+            }
+        }
+
+        private class DataTypeWithCustomEditFormat : DataTypeAttribute
+        {
+            public DataTypeWithCustomEditFormat() : base("Custom datatype")
+            {
+                DisplayFormat = new DisplayFormatAttribute
+                {
+                    ApplyFormatInEditMode = true,
+                    DataFormatString = "value",
+                };
+            }
+        }
+
+        private void ActionWithoutBindAttribute(User param)
+        {
+        }
+
+        private void ActionWithBindAttribute([Bind(new string[] { "IsAdmin" }, Prefix = "ParameterPrefix")] User param)
+        {
+        }
+
+        public class TypeBasedBinderAttribute : Attribute, IBinderMetadata, IModelNameProvider
```

Comment:
Do we still require ```IBinderMetadata``` looks like we no longer have a BinderMetadata in ModelMetadata.